### ISSUE TITLE
[Post-1.4, Discussion] Standardized additional Tile Location Data systems

### ISSUE THREAD

**Issue by Solxanich** [tModLoader/tModLoader#1326]
### Description
We want to standardize some locational data systems for usage. 

Design thoughts for Large-X-Simple:
1) a token setup which means the same structures could be reused and/or use generics.  if someone requested a Token<bool> type, then it would secretly compress it
2)  making Tile a structure? alloc a memory block with the needed memory and we can add in the necessary bytes for all the added data and the make a indexer the wraps it for vanilla stuff and then a generic tile getter for retrieving the specific structure

Further, as an aside, it should be noted that TileID.Sets should be treated as the equivalent to forgeTags. For example, Draedon's Forge from calamity should belong to ModTile.Sets.DraedonForge[], such that Fargo's can quickly add their conglomerate crafting station to that set to get the functionality. 

### What does this proposal attempt to solve or improve?
Currently, there isn't (a) standard system(s) for how data is added for tiles that is locational dependent.
This feature request aims to solve that. Benefits range from optimized performance with well-developed standardized systems, and increased accessibility for modders.

At time of creation, we've tagged areas of consideration to understand what is covered:
[background]
1) Small-Sparse-Simple at one end (1 tile actively using data in runtime, sparse locations to save/load, has simple data) 
2) At other end, large-populous-complex (many tiles actively using data in runtime, large [possibly adjacent] quantities to save/load, and has complex data)

This resulted in the following coverage:
1) small-sparse-complex: TileEntity.      
a) Terraria examples: Terraria Calamity charging station
b) Minecraft reference: Thermal machines
2) small-sparse-simple: Not covered in 1.4 branch.     
a) Terraria examples: Modloader UnloadedTiles, Ancients Awakened sky altar
b) Minecraft reference: Tinker's crafting station with items staying on it
3) Large-X-simple: informally covered by DataCore
a) Terraria examples: Overhaul's fire system
b) Minecraft reference: N/A

### Which (other) solutions should be considered?
This issue is discussion oriented, and should be open to multiple options. Originates from a discord discussion.

**Comment by Rartrin** [comment]
Another solution building on the idea I had for Medium-X-Simple.
If we make tile be a structure, then we can store an array inside of each which will handle all of the data. Effectively, all existing tile data will become part of the new system but we can keep properties for vanilla stuff. This would be Large-X-Simple as well. I'm not sure if it would be better than DataCore, or unmanaged memory allocation, but it would be another idea to explore.

**Comment by Solxanich** [comment]
tl;dr Perhaps the appropriate solution for Large-X-Simple's data size issue is to first check the max expected size in bytes of the added data:
1) if every tile would have its data contained within 2 bytes then just save into the tile struct the array
2) else, store a ushort dataID into the tile struct, where dataID is its de-duplicated data index in a list of struct data 

I'm also thinking that Small-Sparse-Simple could likely be positionIndexed system, written in the DataCore format, as it makes sense due to impacts of sparsity and limited rendering needs.

-----------------------------------------------------------
Finishing looking into DataCore and comparing to positionIndexed system used in unloadedRework yields they are mostly similar.
Two main points of note in memory optimization in positionIndexed system are that:
1) It uses a single int32 instead of point16 for lookup by calculating a `positionIndex = x * maxTilesY + y` which saves a little on the overhead of point16. Decomposition is just remainder and division by maxTilesY. 
2) For added data larger than 8 bytes with little variability in the data, it was significantly more efficient for memory usage to use List<struct(int posID, ushort dataID)> so you can de-duplicate data and skip empty positions, and in the case of unloadedRework you could round down as well on posID.

**Comment by Rartrin** [comment]
For the sake of memory, lists are not the best. They are a reference to a reference and have other data with them. Also the amount of memory saved would vary on the order things were loaded and how often those things were used. Honestly, it is probably best to just declare what they will need when the world initializes so we don't need to worry about bounds checking and resizing.

**Comment by Rartrin** [comment]
Right now I'm not too worried about the implementation actually. I have a number of ways to do it. The main concern is determining how modders should access the data in a way that feel natural and makes sense.

For individual bits, if we want optimizations for them, we will probably always need dedicated `GetFlag(Token)` and `SetFlag(Token)` methods.

Some ideas for the general cases though:

Using generics:
`ref CustomData data = tiles[x,y].Get<CustomData>();`
This would be inline with a lot of things. Though no reusability when it comes to structures.

Tokens would allow for reuse of common types:
`ref CustomData data = tiles[x,y].Get(DataToken);`
Would provide a similar clean look as generics and could even have optimizations for the bools included to almost blend with them, but it is less efficient due to passing in extra information. Also more work for the modder to set up.

Another approach would be making the aspects of the data accessed independently:

Custom grids (this is similar to DataCore 1.0)
`CustomTileGrid grid tiles.GetGrid<CustomTileGrid>();`
`grid[x,y].CoolProperty = true;`
Or the straight forward.
`ref CustomData data = tiles.Get<CustomData>(x, y);`
Both of these are simpler but they limit what elements are available at a time so it would mean lots of repeated similar calls with the same x,y but different generics so I would discourage it.

**Comment by Solxanich** [comment]
From Discord. Okay, so we know the size via IL opcode sizeOf when the struct registers itself to be in the moddedInfo array in 'tiles' struct, we toss an exception if the size is too big (lets say 8 bytes for each struct), we retrieve the struct by its pointer at its location based on x * YmaxTiles + Y with a standard Get(<type>) method.
Where we do minimal work to track where the <type> is within the moddedInfo array so that Get(<>) works
I think that summarizes correctly?
Although I guess it wouldn't be an array for moddedInfo, but that's just me typing an easy to type term.

Saving/loading would be still generally the same in TileIO, except based on tiles struct instead of Main.tile.property. For the mod added data, save as simple unmanaged byte array for simplicity  

Finally, consider adding a UnmanagedMemoryType for tML contributor friendliness.
